### PR TITLE
[CommonOperatorLoader, Interpreter Backend] Add support for GatherND Operator

### PR DESCRIPTION
**Pull request by vbaddi**
Summary:
Adding support for GatherND operator,
Also, Added Interpreter Backend Implementation

Documentation: None

Test Plan: Ninja test is run, also Added OnnxImporterTest and InterpreterOperatorTest cases

### CODE REVIEW

**Comment by vbaddi** [comment]
@jfix71 could you please review this change?

**Comment by vbaddi** [comment]
@jfix71 could you please review the changes and provide the comments?

**Comment by vbaddi** [comment]
@jfix71, could you please review this change?

**Comment by jfix71** [comment]
Really sorry for the major delays here, I've been super swamped. I will definitely get to reviewing next week.

File: lib/Backends/Interpreter/InterpreterNodes.cpp [hunk 1]
```diff
@@ -1962,6 +1962,67 @@ void BoundInterpreterFunction::fwdGatherInst(const glow::GatherInst *I) {
...
+    const glow::GatherNDInst *I) {
+
+  Tensor *dataT = getTensor(I->getData());
+  auto &dataTy = dataT->getType();
+  Tensor *indicesT = getTensor(I->getIndices());
+  Tensor *outT = getTensor(I->getDest());
+  auto &indiceTy = indicesT->getType();
+
+  // Get the last dimension of indices Tensor
+  int64_t lastIndicesDimension = indiceTy.dims()[indiceTy.dims().size() - 1];
```
Comment (jfix71) [inline lib/Backends/Interpreter/InterpreterNodes.cpp:1976]: nit: `const`

File: lib/Backends/Interpreter/InterpreterNodes.cpp [hunk 2]
```diff
@@ -1962,6 +1962,67 @@ void BoundInterpreterFunction::fwdGatherInst(const glow::GatherInst *I) {
...
 
+template <typename ElemTy>
+void BoundInterpreterFunction::fwdGatherNDInstImpl(
+    const glow::GatherNDInst *I) {
+
+  Tensor *dataT = getTensor(I->getData());
+  auto &dataTy = dataT->getType();
+  Tensor *indicesT = getTensor(I->getIndices());
+  Tensor *outT = getTensor(I->getDest());
+  auto &indiceTy = indicesT->getType();
```
Comment (jfix71) [inline lib/Backends/Interpreter/InterpreterNodes.cpp:1973]: nit: `indiceTy` -> `indicesTy`

File: tests/unittests/OnnxImporterTest.cpp [hunk 3]
```diff
@@ -2238,30 +2238,57 @@ TEST_F(OnnxImporterTest, expandDims) {
...
-/// Test loading Gather from an ONNX model.
-TEST_F(OnnxImporterTest, gather) {
-  ExecutionEngine EE;
+/// Helper method to run the gather operator test cases.
+/// \p filename contains the model .onnxtxt.
+/// \p dataShape: data Tensor dimensions.
+/// \p indicesShape: indices Tensor dimensions
+/// \p expectedValues : output Tensor values expected.
+template <class OpType>
+static void gatherTestHelper(std::string fileName,
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp:2247]: nit: const ref, or llvm::StringRef

File: lib/Backends/Interpreter/Interpreter.cpp [hunk 4]
```diff
@@ -470,11 +470,14 @@ bool Interpreter::isOpSupported(const NodeInfo &NI) const {
 
   case Kinded::Kind::GatherNodeKind:
     // Note: Data and Result can be any data type, but must match.
-    return (NI.getInElemTy(GatherNode::DataIdx) ==
-            NI.getOutElemTy(GatherNode::ResultIdx)) &&
```
Comment (jfix71) [inline lib/Backends/Interpreter/Interpreter.cpp:474]: Why remove the requirement here?

**Review by jfix71 — COMMENTED** [review]
@vbaddi Thanks for adding this! Question: GatherND is pretty similar to Gather, of course. I'm wondering if perhaps we could merge the two, for example by lowering Gather to GatherND? What are your thoughts?

**Comment by vbaddi** [comment]
> @vbaddi Thanks for adding this! Question: GatherND is pretty similar to Gather, of course. I'm wondering if perhaps we could merge the two, for example by lowering Gather to GatherND? What are your thoughts?

Sure.
Will try to work on it.

File: include/glow/Graph/Graph.h [hunk 5]
```diff
@@ -1415,6 +1415,12 @@ class Function final : public IRContainer {
   GatherNode *createGather(llvm::StringRef name, NodeValue data,
                            NodeValue indices, unsigned_t batchDims = 0);
 
+  /// Given data tensor of rank r >= 1, indices tensor of rank q >= 1,
```
[the remark this answers is not in the document]
Comment (vbaddi) [reply include/glow/Graph/Graph.h]: Fixed

[the remark this answers is not in the document]
Comment (vbaddi) [reply include/glow/Graph/Graph.h]: Fixed

File: lib/Backends/Interpreter/InterpreterNodes.cpp (hunk 2 above)
Comment (vbaddi) [reply lib/Backends/Interpreter/InterpreterNodes.cpp]: Fixed

File: tests/unittests/OnnxImporterTest.cpp (hunk 3 above)
Comment (vbaddi) [reply tests/unittests/OnnxImporterTest.cpp]: Fixed

File: lib/Backends/Interpreter/Interpreter.cpp (hunk 4 above)
Comment (vbaddi) [reply lib/Backends/Interpreter/Interpreter.cpp]: Rebase issue.
Fixed now.

**Comment by vbaddi** [comment]
@jfix71 Thank you so much for review.
Can this be merged now?

File: lib/Backends/Interpreter/InterpreterNodes.cpp [hunk 6]
```diff
@@ -1962,6 +1962,68 @@ void BoundInterpreterFunction::fwdGatherInst(const glow::GatherInst *I) {
...
+  auto &dataTy = dataT->getType();
+  Tensor *indicesT = getTensor(I->getIndices());
+  Tensor *outT = getTensor(I->getDest());
+  auto &indicesTy = indicesT->getType();
+
+  // Get the last dimension of indices Tensor
+  const int64_t lastIndicesDimension =
+      indicesTy.dims()[indicesTy.dims().size() - 1];
+
+  size_t out_p = 0;
```
Comment (jfix71) [inline lib/Backends/Interpreter/InterpreterNodes.cpp:1979]: nit: 
```suggestion
  size_t outP = 0;
```

**Review by jfix71 — APPROVED** [review]
LGTM -- except seems like you need to add at least `GatherDataNDFloat16IdxInt32` to the OpenCLOperatorTest blacklist. And one nit. Sorry for the delays, please tag me once you make the change and I will get this landed.

File: lib/Backends/Interpreter/InterpreterNodes.cpp (hunk 6 above)
Comment (vbaddi) [reply lib/Backends/Interpreter/InterpreterNodes.cpp]: Done.
@jfix71 Can we land this now?
Thank you

**Comment by jfix71** [comment]
Thanks -- I will fix the conflict on my end before merging.

**Pull request closed without merging**